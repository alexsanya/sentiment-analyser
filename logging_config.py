"""
Logging configuration for news-powered trading system.
Provides structured logging with file separation by log level.
"""

import logging
import logging.handlers
import os
from typing import Any, Dict

import structlog


def setup_logging(environment: str = "development") -> structlog.stdlib.BoundLogger:
    """
    Configure structured logging with file separation by log level.
    
    Args:
        environment: "development" or "production"
        
    Returns:
        Configured structlog logger
    """
    # Ensure logs directory exists
    logs_dir = "logs"
    os.makedirs(logs_dir, exist_ok=True)
    
    # Configure standard library logging
    logging.basicConfig(
        format="%(message)s",
        stream=None,  # We'll handle streams via handlers
        level=logging.INFO,
    )
    
    # Create file handlers for different log levels
    handlers = []
    
    # Error log handler
    error_handler = logging.handlers.RotatingFileHandler(
        filename=os.path.join(logs_dir, "error.log"),
        maxBytes=10 * 1024 * 1024,  # 10MB
        backupCount=5
    )
    error_handler.setLevel(logging.ERROR)
    handlers.append(error_handler)
    
    # Warning log handler
    warning_handler = logging.handlers.RotatingFileHandler(
        filename=os.path.join(logs_dir, "warning.log"),
        maxBytes=10 * 1024 * 1024,  # 10MB
        backupCount=5
    )
    warning_handler.setLevel(logging.WARNING)
    handlers.append(warning_handler)
    
    # Info log handler (all levels)
    info_handler = logging.handlers.RotatingFileHandler(
        filename=os.path.join(logs_dir, "app.log"),
        maxBytes=10 * 1024 * 1024,  # 10MB
        backupCount=5
    )
    info_handler.setLevel(logging.INFO)
    handlers.append(info_handler)
    
    # Console handler for development
    if environment == "development":
        console_handler = logging.StreamHandler()
        console_handler.setLevel(logging.INFO)
        handlers.append(console_handler)
    
    # Configure root logger
    root_logger = logging.getLogger()
    root_logger.handlers.clear()
    for handler in handlers:
        root_logger.addHandler(handler)
    
    # Configure structlog
    processors = [
        structlog.stdlib.filter_by_level,
        structlog.stdlib.add_logger_name,
        structlog.stdlib.add_log_level,
        structlog.stdlib.PositionalArgumentsFormatter(),
        structlog.processors.TimeStamper(fmt="ISO"),
        structlog.processors.StackInfoRenderer(),
        structlog.processors.format_exc_info,
    ]
    
    if environment == "production":
        # JSON formatter for production
        processors.append(structlog.processors.JSONRenderer())
    else:
        # Pretty formatter for development
        processors.append(structlog.dev.ConsoleRenderer())
    
    structlog.configure(
        processors=processors,
        wrapper_class=structlog.stdlib.BoundLogger,
        logger_factory=structlog.stdlib.LoggerFactory(),
        context_class=dict,
        cache_logger_on_first_use=True,
    )
    
    return structlog.get_logger()


def get_logger(name: str = None) -> structlog.stdlib.BoundLogger:
    """
    Get a configured logger instance.
    
    Args:
        name: Logger name (typically __name__)
        
    Returns:
        Configured structlog logger
    """
    return structlog.get_logger(name)