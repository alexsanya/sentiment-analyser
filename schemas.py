"""Shared schema definitions for message validation."""

from typing import List, Optional, Union
from pydantic import BaseModel, Field, field_validator


class DataSource(BaseModel):
    """Schema for data source"""
    name: str = Field("", description="Data source name")
    author_name: str = Field("", description="Data source author")
    author_id: str = Field("", description="Data source author id")

class TweetOutput(BaseModel):
    """Schema for transformed tweet data output"""
    data_source: DataSource = Field(default_factory=DataSource, description="Data source")
    createdAt: int = Field(0, description="Tweet creation datetime as unix timestamp")
    text: str = Field("", description="Tweet text content")
    media: List[str] = Field(default_factory=list, description="Media URLs from tweet")
    links: List[str] = Field(default_factory=list, description="External links from tweet")
    
    @field_validator('text')
    @classmethod
    def validate_text(cls, v):
        if not isinstance(v, str):
            return ""
        return v
    
    @field_validator('media', 'links')
    @classmethod
    def validate_url_lists(cls, v):
        if not isinstance(v, list):
            return []
        return [url for url in v if isinstance(url, str) and url.strip()]