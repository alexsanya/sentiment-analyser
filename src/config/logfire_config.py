"""Configuration for Logfire observability and instrumentation."""

import os
import logfire
from typing import Optional
from ..config.logging_config import get_logger

logger = get_logger(__name__)


def initialize_logfire() -> None:
    """
    Initialize Logfire with configuration from environment variables.
    
    Environment variables:
        LOGFIRE_TOKEN: The Logfire API token (required)
        LOGFIRE_SERVICE_NAME: Name of the service for tracing (default: tweets-notifier)
        LOGFIRE_ENVIRONMENT: Environment name (default: development)
        LOGFIRE_ENABLED: Enable/disable Logfire (default: true)
    """
    # Check if Logfire is enabled
    if os.getenv("LOGFIRE_ENABLED", "true").lower() != "true":
        logger.info("Logfire instrumentation disabled via LOGFIRE_ENABLED")
        return
    
    # Get configuration from environment
    token = os.getenv("LOGFIRE_TOKEN")
    service_name = os.getenv("LOGFIRE_SERVICE_NAME", "sentiment-analyzer")
    environment = os.getenv("LOGFIRE_ENVIRONMENT", os.getenv("ENVIRONMENT", "development"))
    
    if not token:
        logger.warning("LOGFIRE_TOKEN not provided - Logfire instrumentation disabled")
        return
    
    try:
        # Configure Logfire
        logfire.configure(
            token=token,
            #service_name=service_name,
            #service_version=os.getenv("SERVICE_VERSION", "0.1.0"),
            #environment=environment,
        )
        
        # Enable PydanticAI instrumentation
        logfire.instrument_pydantic_ai()
        
        logger.info(
            "Logfire initialized successfully",
            service_name=service_name,
            environment=environment
        )
        
    except Exception as e:
        logger.error("Failed to initialize Logfire", error=str(e))
        # Don't raise - let the application continue without observability


def create_logfire_span(operation: str, **kwargs) -> Optional[logfire.LogfireSpan]:
    """
    Create a Logfire span for tracking operations.
    
    Args:
        operation: The operation name for the span
        **kwargs: Additional attributes to add to the span
        
    Returns:
        Logfire span context manager or None if Logfire is not configured
    """
    try:
        return logfire.span(operation, **kwargs)
    except Exception as e:
        logger.debug("Failed to create Logfire span", operation=operation, error=str(e))
        return None


def log_agent_metrics(
    agent_type: str,
    execution_time: float,
    input_size: int,
    result_type: str,
    success: bool,
    **additional_attributes
) -> None:
    """
    Log agent execution metrics to Logfire.
    
    Args:
        agent_type: Type of agent (text, image, firecrawl)
        execution_time: Time taken for execution in seconds
        input_size: Size of input data (text length, image URL length, etc.)
        result_type: Type of result returned
        success: Whether the operation was successful
        **additional_attributes: Additional attributes to log
    """
    try:
        logfire.info(
            "Agent execution completed",
            agent_type=agent_type,
            execution_time=execution_time,
            input_size=input_size,
            result_type=result_type,
            success=success,
            **additional_attributes
        )
    except Exception as e:
        logger.debug("Failed to log agent metrics", error=str(e))


def is_logfire_enabled() -> bool:
    """Check if Logfire is enabled and configured."""
    return (
        os.getenv("LOGFIRE_ENABLED", "true").lower() == "true" and
        os.getenv("LOGFIRE_TOKEN") is not None
    )