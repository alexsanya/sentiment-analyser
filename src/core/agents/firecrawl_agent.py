"""Firecrawl agent for token detection in web content."""

import time
from typing import cast
from pydantic_ai import Agent
from pydantic_ai.mcp import MCPServerSSE

from ...models.schemas import SentimentAnalysisResult, NoTokenFound
from ...config.logging_config import get_logger
from ...config.sentiment_config import FIRECRAWL_SEARCH_PROMPT, DEFAULT_AGENT_RETRIES
from ...config.logfire_config import create_logfire_span, log_agent_metrics

logger = get_logger(__name__)


class FirecrawlAgent:
    """
    A PydanticAI agent that uses Firecrawl MCP server for web scraping capabilities
    """
    
    def __init__(self, model_name: str = "openai:gpt-4o", firecrawl_url: str = "http://localhost:3000/sse"):
        """
        Initialize the agent with Firecrawl MCP server connection.
        
        Args:
            model_name: The LLM model to use (default: gpt-4o)
            firecrawl_url: The Firecrawl MCP server URL
        """
        # Create MCP server connection to Firecrawl running in SSE mode
        self.firecrawl_server = MCPServerSSE(
            url=firecrawl_url,  # Configurable SSE endpoint for Firecrawl MCP
            tool_prefix='firecrawl'  # Optional: prefix tools to avoid naming conflicts
        )
        
        # Initialize the agent with the MCP server
        self.agent = Agent[None, SentimentAnalysisResult](  # type: ignore[call-overload]
            model=model_name,
            result_type=SentimentAnalysisResult,
            retries=DEFAULT_AGENT_RETRIES,
            system_prompt=FIRECRAWL_SEARCH_PROMPT,
            mcp_servers=[self.firecrawl_server]
        )
    
    async def run(self, url: str) -> SentimentAnalysisResult:
        """
        Process a URL using the agent with Firecrawl capabilities.
        
        Args:
            url: The url to crawl
            
        Returns:
            The token data if found
        """
        start_time = time.time()
        url_length = len(url)
        
        # Create Logfire span for tracing
        span = create_logfire_span(
            "firecrawl_agent.run",
            url=url,
            url_length=url_length,
            agent_type="firecrawl"
        )
        
        try:
            if span:
                with span:
                    async with self.agent.run_mcp_servers():
                        result = await self.agent.run(url)
            else:
                async with self.agent.run_mcp_servers():
                    result = await self.agent.run(url)
            
            execution_time = time.time() - start_time
            result_type = type(result.output).__name__
            
            # Log metrics to Logfire
            log_agent_metrics(
                agent_type="firecrawl",
                execution_time=execution_time,
                input_size=url_length,
                result_type=result_type,
                success=True,
                url=url
            )
            
            logger.info(
                "FirecrawlAgent completed analysis", 
                url=url,
                execution_time=execution_time,
                result_type=result_type
            )
            return cast(SentimentAnalysisResult, result.output)
            
        except Exception as e:
            execution_time = time.time() - start_time
            
            # Log failed metrics to Logfire
            log_agent_metrics(
                agent_type="firecrawl",
                execution_time=execution_time,
                input_size=url_length,
                result_type="NoTokenFound",
                success=False,
                error=str(e),
                url=url
            )
            
            logger.error(
                "FirecrawlAgent failed", 
                error=str(e), 
                url=url,
                execution_time=execution_time
            )
            return NoTokenFound()