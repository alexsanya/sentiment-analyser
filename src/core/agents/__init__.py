"""Sentiment analysis agents for token detection and Trump-Zelenskyy meeting analysis."""

from .text_search_agent import TextSearchAgent
from .image_search_agent import ImageSearchAgent
from .firecrawl_agent import FirecrawlAgent
from .topic_filter_agent import TopicFilterAgent
from .duplicate_detector_agent import DuplicateDetectorAgent
from .geo_expert_agent import GeoExpertAgent
from .retry_wrapper import AgentRetryWrapper

# Keep legacy import for backward compatibility
from .topic_sentiment_agent import TopicSentimentAgent

__all__ = [
    'TextSearchAgent', 
    'ImageSearchAgent', 
    'FirecrawlAgent',
    'TopicFilterAgent',
    'DuplicateDetectorAgent',
    'GeoExpertAgent',
    'TopicSentimentAgent',  # Deprecated but kept for compatibility
    'AgentRetryWrapper'
]