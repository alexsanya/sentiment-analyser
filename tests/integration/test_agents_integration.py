"""Integration tests for PydanticAI agents with real OpenAI API calls.

These tests make real API calls to OpenAI and use snapshot testing to verify
the structured responses from each agent. Requires OPENAI_API_KEY environment variable.
"""

import os
import pytest
import asyncio
from syrupy import SnapshotAssertion

from src.config.logfire_config import initialize_logfire
from src.core.agents.text_search_agent import TextSearchAgent
from src.core.agents.image_search_agent import ImageSearchAgent
from src.core.agents.firecrawl_agent import FirecrawlAgent
from src.models.schemas import SentimentAnalysisResult, TokenDetails
from dotenv import load_dotenv
from src.config.logging_config import setup_logging, get_logger

# Initialize Logfire for integration tests
load_dotenv()
initialize_logfire()

setup_logging("development")  # Auto-detects environment
logger = get_logger(__name__)

from .test_data import (
    TEXT_SAMPLES, 
    TEXT_SAMPLE_DESCRIPTIONS,
    IMAGE_URLS,
    IMAGE_DESCRIPTIONS, 
    WEB_URLS,
    WEB_DESCRIPTIONS
)


def serialize_for_snapshot(result):
    """Serialize agent result for snapshot testing, excluding variable fields.
    
    Excludes the 'definition_fragment' field from TokenDetails to ensure
    stable snapshots since this field contains variable text snippets.
    """
    if hasattr(result, 'model_dump'):
        if isinstance(result, TokenDetails):
            return result.model_dump(exclude={'definition_fragment'})
        return result.model_dump()
    else:
        # Fallback for older Pydantic versions
        if hasattr(result, '__class__') and result.__class__.__name__ == 'TokenDetails':
            return result.dict(exclude={'definition_fragment'})
        return result.dict()


# Mark all tests in this module as integration tests
pytestmark = pytest.mark.integration


class TestEnvironmentSetup:
    """Test environment setup and requirements."""
    
    def test_openai_api_key_required(self):
        """Verify OpenAI API key is available."""
        api_key = os.getenv('OPENAI_API_KEY')
        assert api_key is not None, "OPENAI_API_KEY environment variable is required for integration tests"
        assert len(api_key.strip()) > 0, "OPENAI_API_KEY cannot be empty"


@pytest.fixture
def text_search_agent():
    """Create TextSearchAgent instance."""
    return TextSearchAgent()


@pytest.fixture 
def image_search_agent():
    """Create ImageSearchAgent instance."""
    return ImageSearchAgent()


@pytest.fixture
def firecrawl_agent():
    """Create FirecrawlAgent instance."""
    return FirecrawlAgent()


class TestTextSearchAgent:
    """Integration tests for TextSearchAgent with real OpenAI API calls."""
    
    @pytest.mark.asyncio
    @pytest.mark.parametrize("text_sample,description", zip(TEXT_SAMPLES, TEXT_SAMPLE_DESCRIPTIONS))
    async def test_text_analysis(self, text_search_agent: TextSearchAgent, text_sample: str, description: str, snapshot: SnapshotAssertion):
        """Test TextSearchAgent analysis with various text samples."""
        result = await text_search_agent.run(text_sample)
        
        # Verify result is one of the expected types
        assert isinstance(result, SentimentAnalysisResult.__args__)
        
        # Serialize to dict for snapshot comparison (excluding variable fields)
        result_json = serialize_for_snapshot(result)
            
        # Use description in snapshot name for clarity
        assert result_json == snapshot(name=f"text_search_{description}")


class TestImageSearchAgent:
    """Integration tests for ImageSearchAgent with real OpenAI API calls."""
    
    @pytest.mark.asyncio
    @pytest.mark.parametrize("image_url,description", zip(IMAGE_URLS, IMAGE_DESCRIPTIONS))
    async def test_image_analysis(self, image_search_agent: ImageSearchAgent, image_url: str, description: str, snapshot: SnapshotAssertion):
        """Test ImageSearchAgent analysis with various image URLs."""
        result = await image_search_agent.run(image_url)
        
        # Verify result is one of the expected types
        assert isinstance(result, SentimentAnalysisResult.__args__)
        
        # Serialize to dict for snapshot comparison (excluding variable fields)
        result_json = serialize_for_snapshot(result)
            
        # Use description in snapshot name for clarity
        assert result_json == snapshot(name=f"image_search_{description}")


class TestFirecrawlAgent:
    """Integration tests for FirecrawlAgent with real OpenAI API calls.
    
    Note: These tests require a running Firecrawl MCP server at localhost:3000/sse
    """
    
    @pytest.mark.asyncio
    @pytest.mark.parametrize("web_url,description", zip(WEB_URLS, WEB_DESCRIPTIONS))
    async def test_web_scraping_analysis(self, firecrawl_agent: FirecrawlAgent, web_url: str, description: str, snapshot: SnapshotAssertion):
        """Test FirecrawlAgent analysis with web URLs."""
        result = await firecrawl_agent.run(web_url)
        
        # Verify result is one of the expected types
        assert isinstance(result, SentimentAnalysisResult.__args__)
        
        # Serialize to dict for snapshot comparison (excluding variable fields)
        result_json = serialize_for_snapshot(result)
            
        # Use description in snapshot name for clarity
        assert result_json == snapshot(name=f"firecrawl_{description}")


class TestAgentReliability:
    """Tests for agent reliability and error handling."""
    
    @pytest.mark.asyncio
    async def test_text_agent_empty_input(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test TextSearchAgent with empty input."""
        result = await text_search_agent.run("")
        
        result_json = serialize_for_snapshot(result)
            
        assert result_json == snapshot(name="text_search_empty_input")
    
    @pytest.mark.asyncio
    async def test_text_agent_very_long_input(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test TextSearchAgent with very long input."""
        long_text = "This is a test message. " * 100 + "My new token: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
        result = await text_search_agent.run(long_text)
        
        result_json = serialize_for_snapshot(result)
            
        assert result_json == snapshot(name="text_search_very_long_input")


# Individual test methods for easier debugging
class TestIndividualTextCases:
    """Individual test methods for each text case for easier debugging and snapshot management."""
    
    @pytest.mark.asyncio
    async def test_polygon_explicit_chain_evm_address(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test case 0: Explicit Polygon chain with EVM address."""
        text = TEXT_SAMPLES[0]
        result = await text_search_agent.run(text)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_no_chain_evm_address_infer_ethereum(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test case 1: No chain info, EVM address (should infer Ethereum)."""
        text = TEXT_SAMPLES[1]
        result = await text_search_agent.run(text)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_solana_explicit_chain_solana_address(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test case 2: Explicit Solana chain with Solana address."""
        text = TEXT_SAMPLES[2]
        result = await text_search_agent.run(text)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_no_chain_solana_address_infer_solana(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test case 3: No chain info, Solana address (should infer Solana)."""
        text = TEXT_SAMPLES[3]
        result = await text_search_agent.run(text)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_no_announcement_no_token(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test case 4: No announcement, no token."""
        text = TEXT_SAMPLES[4]
        result = await text_search_agent.run(text)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_purchase_not_release_solana_address(self, text_search_agent: TextSearchAgent, snapshot: SnapshotAssertion):
        """Test case 5: Purchase mention, not release (Solana address)."""
        text = TEXT_SAMPLES[5]
        result = await text_search_agent.run(text)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot


# Additional individual tests for remaining cases...
class TestIndividualImageCases:
    """Individual test methods for each image case."""
    
    @pytest.mark.asyncio
    async def test_trump_token_announcement(self, image_search_agent: ImageSearchAgent, snapshot: SnapshotAssertion):
        """Test Trump token announcement image."""
        image_url = IMAGE_URLS[0]
        result = await image_search_agent.run(image_url)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_melania_token_announcement(self, image_search_agent: ImageSearchAgent, snapshot: SnapshotAssertion):
        """Test Melania token announcement image."""
        image_url = IMAGE_URLS[1]
        result = await image_search_agent.run(image_url)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_non_announcement_image(self, image_search_agent: ImageSearchAgent, snapshot: SnapshotAssertion):
        """Test non-announcement image."""
        image_url = IMAGE_URLS[2]
        result = await image_search_agent.run(image_url)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot


class TestIndividualWebCases:
    """Individual test methods for each web case."""
    
    @pytest.mark.asyncio
    async def test_flockerz_token_website(self, firecrawl_agent: FirecrawlAgent, snapshot: SnapshotAssertion):
        """Test Flockerz token website."""
        web_url = WEB_URLS[0]
        result = await firecrawl_agent.run(web_url)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot
        
    @pytest.mark.asyncio
    async def test_trump_memes_website(self, firecrawl_agent: FirecrawlAgent, snapshot: SnapshotAssertion):
        """Test Trump memes website."""
        web_url = WEB_URLS[1]
        result = await firecrawl_agent.run(web_url)
        result_json = serialize_for_snapshot(result)
        assert result_json == snapshot