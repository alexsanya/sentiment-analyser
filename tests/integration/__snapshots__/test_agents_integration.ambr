# serializer version: 1
# name: TestAgentReliability.test_text_agent_empty_input[text_search_empty_input]
  dict({
  })
# ---
# name: TestAgentReliability.test_text_agent_very_long_input[text_search_very_long_input]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': 1,
    'chain_name': 'Ethereum Mainnet',
    'is_release': True,
    'token_address': '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  })
# ---
# name: TestFirecrawlAgent.test_web_scraping_analysis[https://flockerz.com/-flockerz_token_website][firecrawl_flockerz_token_website]
  dict({
    'chain_defined_explicitly': True,
    'chain_id': None,
    'chain_name': 'Ethereum Mainnet',
    'is_release': True,
    'token_address': '0xb33D999469a7e6b9EbC25A3a05248287b855eD46',
  })
# ---
# name: TestFirecrawlAgent.test_web_scraping_analysis[https://gettrumpmemes.com/-trump_memes_website][firecrawl_trump_memes_website]
  dict({
    'chain_defined_explicitly': True,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN',
  })
# ---
# name: TestImageSearchAgent.test_image_analysis[https://bitcoinworld.co.in/wp-content/uploads/Melania-Trumps-Meme-Coin-MELANIA-B.jpg-non_announcement_image][image_search_non_announcement_image]
  dict({
  })
# ---
# name: TestImageSearchAgent.test_image_analysis[https://lh7-rt.googleusercontent.com/docsz/AD_4nXcNy9kHtzbuH-N-F9B8zy7oqkUobYkzXWhAKqI4qXP7JSIihBNHhJJfz-1gmvJDnxYiTPHRinIe8wBQ3VMBZU0aGxyb6U8k6SWGU5NleZg2AVGyxI7WuyJGcUJ73oG_THgfF_bX?key=qxx7aSSAQfwoulEynZFGQMDH-melania_token_announcement][image_search_melania_token_announcement]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': None,
    'is_release': True,
    'token_address': 'FUAfBo2jgks6gB4Z4LfZkqSZgzNucisEHqnNebaRxM1P',
  })
# ---
# name: TestImageSearchAgent.test_image_analysis[https://pbs.twimg.com/media/GhivrlDWAAA7Ex3?format=jpg&name=medium-trump_token_announcement][image_search_trump_token_announcement]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': None,
    'is_release': True,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqj2jfGiPN',
  })
# ---
# name: TestIndividualImageCases.test_melania_token_announcement
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': None,
    'is_release': True,
    'token_address': 'FUAfBo2jgks6gB4Z4LfZkqSZgzNucisEHqnNebaRxM1P',
  })
# ---
# name: TestIndividualImageCases.test_non_announcement_image
  dict({
  })
# ---
# name: TestIndividualImageCases.test_trump_token_announcement
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqj2jfGiPN',
  })
# ---
# name: TestIndividualTextCases.test_no_announcement_no_token
  dict({
  })
# ---
# name: TestIndividualTextCases.test_no_chain_evm_address_infer_ethereum
  dict({
    'chain_defined_explicitly': False,
    'chain_id': 1,
    'chain_name': 'Ethereum Mainnet',
    'is_release': True,
    'token_address': '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  })
# ---
# name: TestIndividualTextCases.test_no_chain_solana_address_infer_solana
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB',
  })
# ---
# name: TestIndividualTextCases.test_polygon_explicit_chain_evm_address
  dict({
    'chain_defined_explicitly': True,
    'chain_id': 137,
    'chain_name': 'Polygon',
    'is_release': True,
    'token_address': '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  })
# ---
# name: TestIndividualTextCases.test_purchase_not_release_solana_address
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': False,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN',
  })
# ---
# name: TestIndividualTextCases.test_solana_explicit_chain_solana_address
  dict({
    'chain_defined_explicitly': True,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB',
  })
# ---
# name: TestIndividualWebCases.test_flockerz_token_website
  dict({
    'chain_defined_explicitly': True,
    'chain_id': 1,
    'chain_name': 'Ethereum Mainnet',
    'is_release': True,
    'token_address': '0xb33d999469a7e6b9ebc25a3a05248287b855ed46',
  })
# ---
# name: TestIndividualWebCases.test_trump_memes_website
  dict({
    'chain_defined_explicitly': True,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[Accumulationg Bonk on ETH mainnet: 0x1151CB3d861920e07a38e03eEAd12C32178567F6-accumulating_explicit_eth_mainnet][text_search_accumulating_explicit_eth_mainnet]
  dict({
    'chain_defined_explicitly': True,
    'chain_id': 1,
    'chain_name': 'Ethereum Mainnet',
    'is_release': True,
    'token_address': '0x1151CB3d861920e07a38e03eEAd12C32178567F6',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[Accumulationg Bonk: 0x1151CB3d861920e07a38e03eEAd12C32178567F6-accumulating_not_release_evm_address][text_search_accumulating_not_release_evm_address]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': 1,
    'chain_name': 'Ethereum Mainnet',
    'is_release': None,
    'token_address': '0x1151CB3d861920e07a38e03eEAd12C32178567F6',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[Join my webinar tomorrow!-no_announcement_no_token][text_search_no_announcement_no_token]
  dict({
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[Just bought a bad of Trump coin (6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN)-purchase_not_release_solana_address][text_search_purchase_not_release_solana_address]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': False,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[Just bought a bad of Trump coin on solana: (6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN)-purchase_explicit_solana_chain][text_search_purchase_explicit_solana_chain]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': None,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[Just bought a bad of Trump coin on solana: (6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN)-purchase_explicit_solana_chain_duplicate][text_search_purchase_explicit_solana_chain_duplicate]
  dict({
    'chain_defined_explicitly': True,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': None,
    'token_address': '6p6xgHyF7AeE6TZkSmFsko444wqoP15icUSqi2jfGiPN',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[My new token is live on Solana Blockchain: Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB. Buy now!-solana_explicit_chain_solana_address][text_search_solana_explicit_chain_solana_address]
  dict({
    'chain_defined_explicitly': True,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[My new token is live: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48. Happpy trading!-no_chain_evm_address_infer_ethereum][text_search_no_chain_evm_address_infer_ethereum]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': 1,
    'chain_name': 'Ethereum Mainnet',
    'is_release': True,
    'token_address': '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[My new token is live: Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB. Buy now!-no_chain_solana_address_infer_solana][text_search_no_chain_solana_address_infer_solana]
  dict({
    'chain_defined_explicitly': False,
    'chain_id': None,
    'chain_name': 'Solana',
    'is_release': True,
    'token_address': 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB',
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[My new token just released - follow the link to buy!-release_announcement_no_details][text_search_release_announcement_no_details]
  dict({
  })
# ---
# name: TestTextSearchAgent.test_text_analysis[My new token on Polygon blockchain: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48. Buy it now!-polygon_explicit_chain_evm_address][text_search_polygon_explicit_chain_evm_address]
  dict({
    'chain_defined_explicitly': True,
    'chain_id': 137,
    'chain_name': 'Polygon',
    'is_release': True,
    'token_address': '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  })
# ---
